### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by fleea**
- Multiple environment types (SingleBuyEnv, MultipleBuyEnvironment)
- Customizable observations
- Customizable rewards functions
- Integration with Stable Baselines3 for RL agent implementation
- MLflow integration for experiment tracking
- Utility functions for mock tick data generation
- Extensive test suite for environment components

- **New Features**
	- Introduced multiple trading environments tailored for various strategies, including single and multiple buy orders.
	- Added new observation classes to analyze market trends based on historical data.
	- Developed functionality for logging and tracking metrics during training using MLflow.
	- Enhanced reward systems to calculate based on equity changes and price movements.
	- Added utility functions for dynamic environment registration and management.

- **Documentation**
	- Expanded `README.md` to include detailed project features, installation instructions, project structure, and contribution guidelines.

- **Tests**
	- Created extensive unit tests for newly developed environments, observations, and utility functions to ensure robustness and reliability.

- **Chores**
	- Updated dependencies in `requirements.in` and `requirements.txt` for consistency and functionality.
	- Improved the `.gitignore` file to prevent unnecessary files from being tracked.

## Summary by CodeRabbit

- **New Features**
	- Introduced multiple trading environments tailored for various strategies, including single and multiple buy orders.
	- Added new observation classes to analyze market trends based on historical data.
	- Developed functionality for logging and tracking metrics during training using MLflow.
	- Enhanced reward systems to calculate based on equity changes and price movements.
	- Added utility functions for dynamic environment registration and management.

- **Documentation**
	- Expanded `README.md` to include detailed project features, installation instructions, project structure, and contribution guidelines.

- **Tests**
	- Created extensive unit tests for newly developed environments, observations, and utility functions to ensure robustness and reliability.

- **Chores**
	- Updated dependencies in `requirements.in` and `requirements.txt` for consistency and functionality.
	- Improved the `.gitignore` file to prevent unnecessary files from being tracked.

### CODE REVIEW

File: src/environments/single_buy_discrete/single_buy_discrete_env.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    def _get_observation(self: Self) -> spaces:
+        return self.observation.get_observation(self)
+
+    def get_current_price(
+        self, order_action: OrderAction, order_type: Optional[OrderType] = None
+    ) -> float:
+        current = self._get_tick_data()
+        return (
+            current.ask_price if order_action == OrderAction.OPEN else current.bid_price
+        )
```
[the remark this answers is not in the document]
Comment (fleea) [reply src/environments/single_buy_discrete/single_buy_discrete_env.py]: Rejected, if I remove it here: signature of method 'SingleBuyDiscreteEnv.get_current_price()' does not match signature of the base method in class 'BaseEnvironment'

File: src/environments/base_environment/base_environment.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
+from abc import abstractmethod
+from dataclasses import field
+from typing import List, Optional, Any, TypeVar, Tuple, Self
+
+import numpy as np
+import pandas as pd
+from gymnasium import Env, spaces
+from numpy.typing import NDArray
+from decimal import Decimal, getcontext
+
```
[the remark this answers is not in the document]
Comment (fleea) [reply src/environments/base_environment/base_environment.py:10]: rejected. Decimal is used extensively here.

File: src/agents/multiple_buy_gear_agent.py
Comment (dnyoto) [inline src/agents/multiple_buy_gear_agent.py:1]: Find a better name for this environment. E.g. multi-buy-agent

Comment (fleea) [reply src/agents/multiple_buy_gear_agent.py:1]: Done

**Pull request merged**